packages/server/postgres/migrations/2025-04-17T13:33:36.200Z_RBAC.ts

```diff
@@ -0,0 +1,261 @@
+        col.references('Page.id').onDelete('cascade').notNull()
+      )
+      .addColumn('userId', 'varchar(100)', (col) =>
+        col.references('User.id').onDelete('cascade').notNull()
+      )
+      .addColumn('role', sql`"PageRoleEnum"`, (col) => col.notNull())
+      .execute()
+  ])
+
+  await Promise.all([
+    db.schema
+      .createIndex('idx_PageUserAccess_userId')
+      .on('PageUserAccess')
+      .column('userId')
+      .execute(),
+    db.schema
+      .createIndex('idx_PageUserAccess_pageId')
+      .on('PageUserAccess')
+      .column('pageId')
+      .execute(),
+    db.schema
+      .createIndex('idx_PageTeamAccess_teamId')
+      .on('PageTeamAccess')
+      .column('teamId')
+      .execute(),
+    db.schema
+      .createIndex('idx_PageTeamAccess_pageId')
+      .on('PageTeamAccess')
+      .column('pageId')
+      .execute(),
+    db.schema
+      .createIndex('idx_PageOrganizationAccess_orgId')
+      .on('PageOrganizationAccess')
+      .column('orgId')
+      .execute(),
+    db.schema
+      .createIndex('idx_PageOrganizationAccess_pageId')
+      .on('PageOrganizationAccess')
+      .column('pageId')
+      .execute(),
+    db.schema.createIndex('idx_PageAccess_userId').on('PageAccess').column('userId').execute()
+  ])
+
+  await sql`
+-- FUNCTION TO UPDATE CACHED TABLE
+CREATE OR REPLACE FUNCTION "updatePageAccess"("_userId" VARCHAR, "_pageId" INT) RETURNS VOID AS $$
+DECLARE
+  "strongestRole" "PageRoleEnum";
+  "currentRole" "PageRoleEnum";
+BEGIN
+  -- Find the strongest applicable role
+  SELECT MIN(role)
+  INTO "strongestRole"
+  FROM (
+    SELECT pua.role FROM "PageUserAccess" pua
+    WHERE pua."userId" = "_userId" AND pua."pageId" = "_pageId"
+    UNION ALL
+    SELECT pta.role
+    FROM "PageTeamAccess" pta
+    JOIN "TeamMember" tm ON pta."teamId" = tm."teamId"
+    WHERE tm."userId" = "_userId" AND pta."pageId" = "_pageId"
+    UNION ALL
+    SELECT poa.role
+    FROM "PageOrganizationAccess" poa
+    JOIN "OrganizationUser" ou ON poa."orgId" = ou."orgId"
+    WHERE ou."userId" = "_userId" AND poa."pageId" = "_pageId"
+  ) AS effective_roles;
+
+  -- Check existing effective role
+  SELECT role INTO "currentRole"
+  FROM "PageAccess"
+  WHERE "userId" = "_userId" AND "pageId" = "_pageId";
+
+  IF "strongestRole" IS NULL THEN
+    -- User lost access
+    DELETE FROM "PageAccess"
+    WHERE "userId" = "_userId" AND "pageId" = "_pageId";
+  ELSIF "currentRole" IS NULL THEN
+    -- New access
+    INSERT INTO "PageAccess" ("userId", "pageId", role)
+    VALUES ("_userId", "_pageId", "strongestRole");
+  ELSIF "strongestRole" <> "currentRole" THEN
+    -- Changed access
+    UPDATE "PageAccess"
+    SET role = "strongestRole"
+    WHERE "userId" = "_userId" AND "pageId" = "_pageId";
+  END IF;
+END;
+$$ LANGUAGE plpgsql;
+
+
+-- UPDATE CACHE FROM PageUserAccess
+CREATE OR REPLACE FUNCTION "updateUserPageAccess"() RETURNS TRIGGER AS $$
+BEGIN
+  PERFORM "updatePageAccess"(NEW."userId", NEW."pageId");
```

Comment:
-1 You're running this also as a delete trigger, `NEW` is not defined in this case, but we need to reference `OLD` instead.